lib/spack/spack/test/config.py

```diff
@@ -1266,3 +1266,63 @@ def test_user_cache_path_is_default_when_env_var_is_empty(working_env):
     os.environ['SPACK_USER_CACHE_PATH'] = ''
     assert os.path.expanduser("~%s.spack" % os.sep) == \
         spack.paths._get_user_cache_path()
+
+
+github_url = "https://github.com/fake/fake/{0}/develop"
+gitlab_url = "https://gitlab.fake.io/user/repo/-/blob/config/defaults"
+
+
+@pytest.mark.parametrize('url,isfile,fail', [
+    ("https://not.a.real.url/", False, True),
+    (github_url.format('tree'), False, True),
+    ("{0}/etc/fake/defaults".format(github_url.format('tree')), False, False),
+    ("{0}/README.md".format(github_url.format('blob')), True, True),
+    ("{0}/etc/fake/defaults/packages.yaml".format(github_url.format('blob')),
+     True, False),
+    (gitlab_url, False, False),
+    ("{0}/compilers.yaml".format(gitlab_url), True, False),
+    (None, False, True)
+])
+def test_config_fetch_remote_configs(
+        tmpdir, mutable_empty_config, mock_url_config, url, isfile, fail):
+    dest_dir = join_path(tmpdir.strpath, 'defaults')
+    if fail:
+        msg = "Cannot retrieve configuration"
+        with spack.config.override('config:url_fetch_method', 'curl'):
+            with pytest.raises(spack.config.ConfigFileError, match=msg):
+                spack.config.fetch_remote_configs(url, dest_dir)
+    else:
+        with spack.config.override('config:url_fetch_method', 'curl'):
+            path = spack.config.fetch_remote_configs(url, dest_dir)
+            assert path
+            assert os.path.exists(path)
+            if isfile:
+                assert os.path.basename(path) == os.path.basename(url)
+
+
+def test_config_file_dir_failure(tmpdir, mutable_empty_config):
+    with pytest.raises(spack.config.ConfigFileError, match="not a file"):
+        spack.config.read_config_file(tmpdir.strpath)
+
+
+@pytest.mark.skipif(sys.platform == 'win32',
+                    reason="chmod not supported on Windows")
+def test_config_file_perms_failure(tmpdir, mutable_empty_config):
+    # Try a file that cannot be written
+    filename = join_path(tmpdir.strpath, 'test.yaml')
+    touch(filename)
+    os.chmod(filename, 0o200)
+
+    with pytest.raises(spack.config.ConfigFileError, match="not readable"):
+        spack.config.read_config_file(filename)
```

Comment:
I think this test is already covered by `test_bad_command_line_scopes`.

Reply:
I don't think so.  `test_bad_command_line_scopes` checks the path from `_add_command_line_scopes` where it raises the exception; whereas, this test is checking the path in `read_config_file` that will raise a `ConfigFileError`.  This appears to me to be separate as it is called from `add_from_file`, which is called directly by `spack config add`.

Reply:
Renamed and added docstring.

Reply:
Good point: `add_command_line_scopes` has a separate check for permissions. That seems strange to me: I think the test you added here is better and removes the need for `test_bad_command_line_scopes` (since that function ends up calling `read_config_file`).

Reply:
I am not seeing overlap between the two test functions.  Adding a scope does not result in the read, retrieving does.

Reply:
I think this can be resolved on account of https://github.com/spack/spack/pull/29026#issuecomment-1162497381 - thanks for explaining that!